File: src/core/analyses/tran.js
/**
 * Transient analysis.
 *
 * Written as a resumable stepper rather than a run-to-completion loop. The
 * interactive UI needs to advance the simulation for a slice of wall-clock time,
 * hand control back so the browser can paint, and resume where it left off. A
 * blocking loop would have to be rewritten to support that; this does not.
 *
 *   const run = new TransientRun(circuit, { tstop: 1e-3, tstep: 1e-6 });
 *   run.begin();
 *   while (!run.done) run.advanceFor(8);   // 8 ms of wall clock per call
 */
import { Mode } from '../device.js';
import { setIntegrator, lteTimestep, Method } from '../integrator.js';
import { solveDC, newton, ConvergenceError } from '../newton.js';
import { waveformBreakpoints } from '../devices/sources.js';

export class TransientRun {
  /**
   * @param {import('../circuit.js').Circuit} circuit
   * @param {{tstop:number, tstep:number, tstart?:number, tmax?:number,
   *          uic?:boolean, method?:string, onPoint?:Function}} opts
   */
  constructor(circuit, opts) {
    this.c = circuit;
    this.tstop = opts.tstop;
    this.tstep = opts.tstep;
    this.tstart = opts.tstart ?? 0;
    this.tmax = opts.tmax ?? Math.min(this.tstep, (this.tstop - this.tstart) / 50);
    this.uic = opts.uic ?? false;
    this.method = opts.method ?? circuit.options.method;
    this.onPoint = opts.onPoint ?? null;

    this.time = 0;
    this.dt = 0;
    this.done = false;
    this.order = 1;
    this.accepted = 0;
    this.rejected = 0;
    this.iterations = 0;

    /** @type {number[]} times of accepted points */
    this.times = [];
    /** @type {Float64Array[]} solutions at accepted points */
    this.data = [];

    this.breakpoints = [];
    this.bpIndex = 0;
  }

  begin() {
    const c = this.c;
    c.ensureBuilt();
    const ctx = c.ctx;
    ctx.tstep = this.tstep;
    ctx.tstop = this.tstop;

    // Collect waveform breakpoints so no source edge is stepped over.
    const bp = new Set([0, this.tstop]);
    for (const d of c.devices) {
      if (d.tran) for (const t of waveformBreakpoints(d.tran, this.tstop)) bp.add(t);
    }
    this.breakpoints = [...bp].sort((a, b) => a - b);
    this.bpIndex = 0;

    ctx.state.reset();
    if (this.uic) {
      ctx.x.fill(0);
      this._applyInitialConditions();
    } else {
      solveDC(c);
    }
    ctx.xOld.set(ctx.x);

    // First timepoint: charges are seeded from the operating point.
    ctx.mode = Mode.TRAN;
    ctx.time = 0;
    ctx.initTransient = true;
    setIntegrator(ctx, this.method, 1);
    c.load(Mode.TRAN);
    ctx.initTransient = false;

    this.time = 0;
    this.dt = Math.min(this.tmax, this.tstep) / 10;
    this.order = 1;
    this._record(0);
    return this;
  }

  _applyInitialConditions() {
    const ctx = this.c.ctx;
    for (const d of this.c.devices) {
      if (d.ic != null && d.nodes.length >= 2) {
        const [p, n] = d.nodes;
        if (p >= 0) ctx.x[p] = d.ic + (n >= 0 ? ctx.x[n] : 0);
      }
    }
  }

  _record(t) {
    this.times.push(t);
    const snap = Float64Array.from(this.c.ctx.x);
    this.data.push(snap);
    if (this.onPoint) this.onPoint(t, snap);
  }

  /** Largest step allowed by breakpoints, tmax, and the remaining span. */
  _limitStep(dt) {
    let lim = Math.min(dt, this.tmax, this.tstop - this.time);
    while (this.bpIndex < this.breakpoints.length &&
           this.breakpoints[this.bpIndex] <= this.time + 1e-15) {
      this.bpIndex++;
    }
    if (this.bpIndex < this.breakpoints.length) {
      const next = this.breakpoints[this.bpIndex];
      const toBp = next - this.time;
      if (toBp > 0 && toBp < lim) lim = toBp;
      // Do not leave a sliver of a step just before a breakpoint.
      else if (toBp > 0 && toBp < 1.5 * lim) lim = toBp / 2;
    }
    return Math.max(lim, 1e-18);
  }

  /**
   * Attempt one timestep. Returns true if it was accepted.
   */
  step() {
    const c = this.c;
    const ctx = c.ctx;
    const dt = this._limitStep(this.dt);

    ctx.state.advance();
    ctx.dt[1] = ctx.dt[0];
    ctx.dt[0] = dt;
    ctx.time = this.time + dt;
    ctx.xOld.set(ctx.x);

    const order = Math.min(this.order, c.options.maxord);
    setIntegrator(ctx, this.method, this.accepted < 2 ? 1 : order);

    ctx.forceInit = false;
    const r = newton(c, Mode.TRAN, c.options.itl4);
    this.iterations += r.iterations;

    if (!r.converged) {
      ctx.state.rollback();
      ctx.x.set(ctx.xOld);
      ctx.dt[0] = ctx.dt[1];
      this.rejected++;
      this.dt = dt / 8;
      if (this.dt < 1e-18) {
        throw new ConvergenceError(
          `transient failed to converge at t=${this.time.toExponential(4)} s ` +
          `(timestep collapsed to ${this.dt.toExponential(2)} s)`,
          { time: this.time, reason: r.reason }
        );
      }
      return false;
    }

    // Commit charges at the CONVERGED solution.
    //
    // Devices compute q(v) during load(), which runs before the solve. For a
    // nonlinear circuit Newton iterates until v stops moving, so the stored
    // charge ends up consistent. For a linear circuit Newton exits after one
    // iteration and the stored charge is left one timepoint stale, which
    // silently corrupts every subsequent integration and LTE estimate. One
    // extra stamp pass at the converged solution fixes both cases.
    c.load(Mode.TRAN);

    // Local truncation error check across every charge-storing device.
    let allowed = Infinity;
    for (const d of c.devices) {
      if (d.nStates >= 2) {
        for (let k = 0; k < d.nStates; k += 2) {
          const lim = lteTimestep(ctx, d.stateOff + k, this.method, c.options.trtol);
          if (lim < allowed) allowed = lim;
        }
      }
    }

    if (this.accepted >= 2 && allowed < 0.9 * dt) {
      ctx.state.rollback();
      ctx.x.set(ctx.xOld);
      ctx.dt[0] = ctx.dt[1];
      this.rejected++;
      this.dt = Math.max(allowed, dt / 8);
      return false;
    }

    this.time += dt;
    this.accepted++;
    this.order = 2;
    if (this.time >= this.tstart) this._record(this.time);

    // Grow the step, but never by more than 2x, and never past what LTE allows.
    let next = Math.min(2 * dt, allowed);
    if (r.iterations > c.options.itl4 * 0.7) next = Math.min(next, dt);
    this.dt = next;

    if (this.time >= this.tstop - 1e-15) this.done = true;
    return true;
  }

  /** Run steps until `ms` of wall-clock time has elapsed or the run finishes. */
  advanceFor(ms) {
    const t0 = (typeof performance !== 'undefined' ? performance : Date).now();
    while (!this.done) {
      this.step();
      if (((typeof performance !== 'undefined' ? performance : Date).now() - t0) >= ms) break;
    }
    return this.done;
  }

  /** Run the whole analysis. */
  runToCompletion() {
    while (!this.done) this.step();
    return this.result();
  }

  result() {
    return {
      time: this.times,
      data: this.data,
      labels: this.c.labels.slice(),
      stats: {
        accepted: this.accepted,
        rejected: this.rejected,
        iterations: this.iterations,
      },
    };
  }
}

/** Convenience wrapper for a blocking transient run. */
export function tran(circuit, opts) {
  return new TransientRun(circuit, opts).begin().runToCompletion();
}

File: crates/spicelab-core/src/analyses/tran.rs
//! Transient analysis. Port of `src/core/analyses/tran.js`.
//!
//! Written as a resumable stepper rather than a run-to-completion loop. The
//! interactive UI needs to advance the simulation for a slice of wall-clock
//! time, hand control back so the browser can paint, and resume where it left
//! off. A blocking loop would have to be rewritten to support that; this does
//! not. `advance_steps` is the wasm-friendly form of the JS `advanceFor(ms)` —
//! the host decides the budget, since `Instant::now` is not available on
//! `wasm32-unknown-unknown`.

use crate::circuit::Circuit;
use crate::context::{lte_timestep, Method, Mode};
use crate::device::DeviceKind;
use crate::newton::newton;
use crate::newton::solve_dc;
use crate::SimError;

#[derive(Clone, Debug)]
pub struct TranOptions {
    pub tstop: f64,
    pub tstep: f64,
    pub tstart: f64,
    pub tmax: Option<f64>,
    pub uic: bool,
    pub method: Option<Method>,
}

impl TranOptions {
    pub fn new(tstop: f64, tstep: f64) -> Self {
        TranOptions {
            tstop,
            tstep,
            tstart: 0.0,
            tmax: None,
            uic: false,
            method: None,
        }
    }
}

pub struct TranStats {
    pub accepted: usize,
    pub rejected: usize,
    pub iterations: usize,
}

pub struct TranResult {
    pub time: Vec<f64>,
    pub data: Vec<Vec<f64>>,
    pub labels: Vec<String>,
    pub stats: TranStats,
}

pub struct TransientRun {
    tstop: f64,
    tstep: f64,
    tstart: f64,
    tmax: f64,
    uic: bool,
    method: Method,

    pub time: f64,
    dt: f64,
    pub done: bool,
    order: usize,
    pub accepted: usize,
    pub rejected: usize,
    pub iterations: usize,

    pub times: Vec<f64>,
    pub data: Vec<Vec<f64>>,

    breakpoints: Vec<f64>,
    bp_index: usize,
    /// The step just accepted ended on a source discontinuity, so the next one
    /// must start at order 1. See `step`.
    bp_restart: bool,
}

impl TransientRun {
    pub fn new(circuit: &Circuit, opts: &TranOptions) -> Self {
        let tmax = opts
            .tmax
            .unwrap_or_else(|| opts.tstep.min((opts.tstop - opts.tstart) / 50.0));
        TransientRun {
            tstop: opts.tstop,
            tstep: opts.tstep,
            tstart: opts.tstart,
            tmax,
            uic: opts.uic,
            method: opts.method.unwrap_or(circuit.options.method),
            time: 0.0,
            dt: 0.0,
            done: false,
            order: 1,
            accepted: 0,
            rejected: 0,
            iterations: 0,
            times: Vec::new(),
            data: Vec::new(),
            breakpoints: Vec::new(),
            bp_index: 0,
            bp_restart: false,
        }
    }

    pub fn begin(&mut self, c: &mut Circuit) -> Result<(), SimError> {
        c.ensure_built()?;
        c.ctx.tstep = self.tstep;
        c.ctx.tstop = self.tstop;
        c.ctx.method = self.method;

        // Collect waveform breakpoints so no source edge is stepped over.
        let mut bp: Vec<f64> = vec![0.0, self.tstop];
        for d in c.devices.iter() {
            if let Some(w) = d.tran_waveform() {
                bp.extend(w.breakpoints(self.tstop));
            }
        }
        bp.sort_by(|a, b| a.partial_cmp(b).unwrap());
        bp.dedup();
        self.breakpoints = bp;
        self.bp_index = 0;

        c.ctx.state.reset();
        if self.uic {
            c.ctx.x.iter_mut().for_each(|v| *v = 0.0);
            self.apply_initial_conditions(c);
        } else {
            // Sources take their t=0 waveform value here, not their DC value.
            c.ctx.time = 0.0;
            c.ctx.tran_op = true;
            let r = solve_dc(c, None);
            c.ctx.tran_op = false;
            r?;
        }
        c.ctx.x_old.copy_from_slice(&c.ctx.x);

        // First timepoint: charges are seeded from the operating point.
        c.ctx.mode = Mode::Tran;
        c.ctx.time = 0.0;
        c.ctx.init_transient = true;
        c.ctx.order = 1;
        c.load(Mode::Tran);
        c.ctx.init_transient = false;

        self.time = 0.0;
        self.dt = self.tmax.min(self.tstep) / 10.0;
        self.order = 1;
        self.record(c, 0.0);
        Ok(())
    }

    fn apply_initial_conditions(&self, c: &mut Circuit) {
        // Collected first so the borrow on `devices` ends before `ctx` is
        // mutated. Same semantics as the JS loop.
        let ics: Vec<(i32, i32, f64)> = c
            .devices
            .iter()
            .filter_map(|d| {
                let ic = match d {
                    DeviceKind::Capacitor(x) => x.ic,
                    DeviceKind::Inductor(x) => x.ic,
                    _ => None,
                }?;
                let n = d.nodes();
                if n.len() >= 2 {
                    Some((n[0], n[1], ic))
                } else {
                    None
                }
            })
            .collect();
        for (p, n, ic) in ics {
            if p >= 0 {
                let base = if n >= 0 { c.ctx.x[n as usize] } else { 0.0 };
                c.ctx.x[p as usize] = ic + base;
            }
        }
    }

    fn record(&mut self, c: &Circuit, t: f64) {
        self.times.push(t);
        self.data.push(c.ctx.x.clone());
    }

    /// Largest step allowed by breakpoints, tmax, and the remaining span.
    fn limit_step(&mut self, dt: f64) -> f64 {
        let mut lim = dt.min(self.tmax).min(self.tstop - self.time);
        while self.bp_index < self.breakpoints.len()
            && self.breakpoints[self.bp_index] <= self.time + 1e-15
        {
            self.bp_index += 1;
        }
        if self.bp_index < self.breakpoints.len() {
            let to_bp = self.breakpoints[self.bp_index] - self.time;
            if to_bp > 0.0 && to_bp < lim {
                lim = to_bp;
            } else if to_bp > 0.0 && to_bp < 1.5 * lim {
                // Do not leave a sliver of a step just before a breakpoint.
                lim = to_bp / 2.0;
            }
        }
        lim.max(1e-18)
    }

    /// Attempt one timestep. Returns whether it was accepted.
    pub fn step(&mut self, c: &mut Circuit) -> Result<bool, SimError> {
        let dt = self.limit_step(self.dt);

        c.ctx.state.advance();
        c.ctx.dt[1] = c.ctx.dt[0];
        c.ctx.dt[0] = dt;
        c.ctx.time = self.time + dt;
        c.ctx.x_old.copy_from_slice(&c.ctx.x);

        c.ctx.method = self.method;
        // Order restarts at 1 for the first steps of the run AND for the first
        // step after every source discontinuity.
        //
        // Trapezoidal integration carries the previous step's branch current
        // forward (`i = (2/dt)*(q0-q1) - i1`). Across a breakpoint that history
        // belongs to the OTHER side of the edge, so the first post-edge step is
        // seeded with a derivative that no longer exists, and in a branch with
        // no resistance to damp it the error alternates in sign and never
        // decays: a gate overlap capacitance driven by a PULSE rings at twice
        // the correct current forever. The mean stays right, which is why this
        // looks like a plausible waveform rather than an error — and driving
        // that ringing into a load resistor flips the MOSFET's drain/source
        // roles every step, which is what finally showed up, as a timestep
        // collapse on a p-channel switch. One backward-Euler step at the edge
        // discards the stale history, which is what SPICE has always done.
        c.ctx.order = if self.accepted < 2 || self.bp_restart {
            1
        } else {
            self.order.min(c.options.maxord)
        };

        c.ctx.force_init = false;
        let itl4 = c.options.itl4;
        let r = newton(c, Mode::Tran, itl4);
        self.iterations += r.iterations;

        if !r.converged {
            c.ctx.state.rollback();
            c.ctx.x.copy_from_slice(&c.ctx.x_old);
            c.ctx.dt[0] = c.ctx.dt[1];
            self.rejected += 1;
            self.dt = dt / 8.0;
            if self.dt < 1e-18 {
                return Err(SimError::Convergence {
                    message: format!(
                        "transient failed to converge at t={:.4e} s (timestep collapsed to {:.2e} s)",
                        self.time, self.dt
                    ),
                    time: Some(self.time),
                });
            }
            return Ok(false);
        }

        // Commit charges at the CONVERGED solution.
        //
        // Devices compute q(v) during load(), which runs before the solve. For a
        // nonlinear circuit Newton iterates until v stops moving, so the stored
        // charge ends up consistent. For a LINEAR circuit Newton exits after one
        // iteration and the stored charge is left one timepoint stale, which
        // silently corrupts every subsequent integration and LTE estimate. This
        // one extra stamp pass at the converged solution fixes both cases.
        // Do not remove it: nothing in the suite fails loudly if you do.
        c.load(Mode::Tran);

        // Local truncation error check across every charge-storing device.
        let mut allowed = f64::INFINITY;
        let trtol = c.options.trtol;
        for d in c.devices.iter() {
            let ns = d.n_lte_states();
            if ns >= 2 {
                let off = d.state_off();
                for k in (0..ns).step_by(2) {
                    let lim = lte_timestep(&c.ctx, off + k, self.method, trtol);
                    if lim < allowed {
                        allowed = lim;
                    }
                }
            }
        }

        if self.accepted >= 2 && allowed < 0.9 * dt {
            c.ctx.state.rollback();
            c.ctx.x.copy_from_slice(&c.ctx.x_old);
            c.ctx.dt[0] = c.ctx.dt[1];
            self.rejected += 1;
            self.dt = allowed.max(dt / 8.0);
            return Ok(false);
        }

        self.time += dt;
        self.accepted += 1;
        self.order = 2;
        self.bp_restart = self.bp_index < self.breakpoints.len()
            && (self.breakpoints[self.bp_index] - self.time).abs() <= 1e-15;
        if self.time >= self.tstart {
            let t = self.time;
            self.record(c, t);
        }

        // Grow the step, but never by more than 2x, and never past LTE.
        let mut next = (2.0 * dt).min(allowed);
        if r.iterations as f64 > c.options.itl4 as f64 * 0.7 {
            next = next.min(dt);
        }
        self.dt = next;

        if self.time >= self.tstop - 1e-15 {
            self.done = true;
        }
        Ok(true)
    }

    /// Run at most `max_steps` step attempts, then yield. The resumable form:
    /// the caller owns the wall-clock budget.
    pub fn advance_steps(&mut self, c: &mut Circuit, max_steps: usize) -> Result<bool, SimError> {
        for _ in 0..max_steps {
            if self.done {
                break;
            }
            self.step(c)?;
        }
        Ok(self.done)
    }

    /// Run to the end and copy out the trajectory. The run stays intact and
    /// inspectable — `times`/`data` remain readable afterwards, which is what
    /// callers that keep the `TransientRun` around expect.
    pub fn run_to_completion(&mut self, c: &mut Circuit) -> Result<TranResult, SimError> {
        while !self.done {
            self.step(c)?;
        }
        Ok(self.result(c))
    }

    /// Run to the end and take ownership of the trajectory, leaving the run
    /// drained. `data` holds one `Vec` per accepted timepoint, so this avoids
    /// duplicating the whole run's allocations when the caller is finished with
    /// the stepper anyway. Used by [`tran`].
    pub fn into_result(mut self, c: &mut Circuit) -> Result<TranResult, SimError> {
        while !self.done {
            self.step(c)?;
        }
        Ok(TranResult {
            time: std::mem::take(&mut self.times),
            data: std::mem::take(&mut self.data),
            labels: c.labels.clone(),
            stats: TranStats {
                accepted: self.accepted,
                rejected: self.rejected,
                iterations: self.iterations,
            },
        })
    }

    pub fn result(&self, c: &Circuit) -> TranResult {
        TranResult {
            time: self.times.clone(),
            data: self.data.clone(),
            labels: c.labels.clone(),
            stats: TranStats {
                accepted: self.accepted,
                rejected: self.rejected,
                iterations: self.iterations,
            },
        }
    }
}

/// Convenience wrapper for a blocking transient run.
pub fn tran(c: &mut Circuit, opts: &TranOptions) -> Result<TranResult, SimError> {
    let mut run = TransientRun::new(c, opts);
    run.begin(c)?;
    run.into_result(c)
}
